[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89882

Summary:
Added two ops:
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_output_fp32
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_relu_output_fp32

corresponding pattern for `linear_with_input_q_dq_qweight_dq_output_fp32` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear_with_input_q_dq_qweight_dq

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @VitalyFedyunin @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

jerryzh168 committed:
[quant] Add fused "q - qlinear - dq" operator

Summary:
att, this is needed by a meta-internal team

Test Plan:
tbd

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_fused_qdq_skip_quant
* torch.ops.quantized.linear_fused_qdq_skip_quant_relu

corresponding pattern for `linear_fused_qdq_skip_quant` would be:
```
input -> q* -> dq* -> linear* ->
    weight -> q* -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_fused_qdq_skip_quant
* torch.ops.quantized.linear_fused_qdq_skip_quant_relu

corresponding pattern for `linear_fused_qdq_skip_quant` would be:
```
input -> q* -> dq* -> linear* ->
    weight -> q* -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

vkuzo reviewed (commented):
the pattern in the PR summary makes sense.

Would it make sense to have the name be something like `input_qint8_dq_weight_qint8_dq_linear`?  The "skip_quant" seems confusing.

vkuzo commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3535,12 +3535,14 @@ class TestQuantizedLinear(TestCase):
            use_multi_dim_input=st.booleans(),
-           use_channelwise=st.booleans())
+           use_channelwise=st.booleans(),
+           use_fused_qdq_skip_quant=st.booleans())
```
this test case is getting really confusing with so many flags, should we just test this separately?  

Lines such as 

```
            if use_fused_qdq_skip_quant:
                Y_q_dq = qlinear(X, X_scale, X_zp, W_prepack)
            else:
                Y_q = qlinear(X_q, W_prepack, Y_scale, Y_zp)
```

are pretty hard to read for someone who wasn't on this PR review

jerryzh168 replied:
yeah makes sense, although this test also is extremely long I feel we shouldn't just copy paste code, how about making a helper function and expose some branches as argument?

jerryzh168 replied:
actually, are you suggesting to copy-paste the code? I don't see other ways to remove this `if` check..

vkuzo replied:
I don't feel too strongly on how.  If it was me, I'd probably copy-pasta, as I rank readability significantly higher than code reuse in this context.  I guess you could also refactor it somehow, or just ignore the feedback and leave as is.  Either way lgtm.

jerryzh168 commented:
> the pattern in the PR summary makes sense.
> 
> Would it make sense to have the name be something like `input_qint8_dq_weight_qint8_dq_linear`? The "skip_quant" seems confusing.

how about `linear_with_input_q_dq_weight_dq`? this would describe the pattern we are matching

jerryzh168 commented:
@vkuzo please take a look at the reply

vkuzo commented:
```
input -> q* -> dq* -> linear* ->
    weight -> q -> dq* /
```

> how about linear_with_input_q_dq_weight_dq? this would describe the pattern we are matching

I think whatever name describes that pattern the best would sound good to me. `linear_with_input_q_dq_weight_dq` sounds like there is only one dq after the weight, is that correct?  I see `weight->q->dq` in your pattern.

jerryzh168 commented:
> ```
> input -> q* -> dq* -> linear* ->
>     weight -> q -> dq* /
> ```
> 
> > how about linear_with_input_q_dq_weight_dq? this would describe the pattern we are matching
> 
> I think whatever name describes that pattern the best would sound good to me. `linear_with_input_q_dq_weight_dq` sounds like there is only one dq after the weight, is that correct? I see `weight->q->dq` in your pattern.

yeah the q after weight is not part of the pattern, the pattern is highlighted with `*`s, maybe I can use `linear_with_input_q_dq_qweight_dq` to be more precise, although in reality the weight is actually packed

kimishpatel commented:
I may be asking out of place question, but what is the motivation to introduce this new op? it seems same as the qlinear, no?

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -748,6 +922,11 @@ TORCH_LIBRARY_IMPL(_quantized, QuantizedCPU, m) {
 }
 
+TORCH_LIBRARY_IMPL(quantized, CPU, m) {
+  m.impl(TORCH_SELECTIVE_NAME("quantized::linear_fused_qdq_skip_quant"), TORCH_FN(QLinearInt8FusedQDQ<false>::run));
```
So is this for qlinear_dynamic?

kimishpatel replied:
Are we decomposing qlinear_dynamic into input_quant->qlinear?

If so note that this is not a quantized linear op or I dont know what to call this op because it outputs fp32 value.

jerryzh168 replied:
this is similar to qlinear_dynamic, except that the scale/zero_point is calculated statically instead of dynamically

jerryzh168 replied:
if by qlinear you mean static quantized linear, then this is not qlinear

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_fused_qdq_skip_quant
* torch.ops.quantized.linear_fused_qdq_skip_quant_relu

corresponding pattern for `linear_fused_qdq_skip_quant` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_fused_qdq_skip_quant
* torch.ops.quantized.linear_fused_qdq_skip_quant_relu

corresponding pattern for `linear_fused_qdq_skip_quant` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented:
TODO: we can refactor quantized linear to use this operator in the future, to reduce code duplication

vkuzo commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -152,6 +152,14 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, bool reduce_range=False) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
+  // Corresponding pattern:
```
can we clearly explain what the asterisk means here?  I understand it based on your PR comments, but it's not clear from the diagram.

can we also add an "after" state?

vkuzo replied:
since this function is also not consistent with other functions in this file, it would be great to add expected input and output dtypes to the docblock.  We should specify that the input is fp32, and we have a packed weight with int8 format packed for xnnpack/qnnpack/fbgemm, and the output is fp32.

vkuzo commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -152,6 +152,14 @@ TORCH_LIBRARY(quantized, m) {
+  // input -> q* -> dq* -> linear* ->
+  //         qweight -> dq* /
+  m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_with_input_q_dq_qweight_dq(Tensor X, float X_scale, int X_zero_point, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
+  // Corresponding pattern:
```
same here

vkuzo commented on aten/src/ATen/native/quantized/PackedParams.h:
```diff
@@ -36,6 +36,30 @@ struct LinearPackedParamsBase : public torch::jit::CustomClassHolder {
 
+  // input is NOT quantized; quantization of input occurs within the function
+  // output is NOT quantized
+  virtual at::Tensor apply_with_input_q_dq_qweight_dq(
```
can we make sure to include the ASCII graphs on every single c++ callsite added in this file to explain what this is doing?  This is not intuitive and I guarantee that most readers will have no idea what this function is doing based on the name, and unfortunately there isn't a way to easily navigate from here to the existing docblocks in aten/src/ATen/native/quantized/library.cpp

jerryzh168 replied:
sure, you mean just the two functions in this file right?

jerryzh168 replied:
sure, will change

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_fused_qdq_skip_quant
* torch.ops.quantized.linear_fused_qdq_skip_quant_relu

corresponding pattern for `linear_fused_qdq_skip_quant` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3671,6 +3672,135 @@ def test_qlinear(self, batch_size, input_channels, output_channels, use_bias,
+            use_relu, use_multi_dim_input, use_channelwise):
+        decimal_val = 4
+        dtypes = [torch.quint8]
+        if torch.backends.quantized.engine == 'qnnpack':
```
it's not, I have a skipIfNoFBGEMM for this test actually, I can remove this code

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_relu

corresponding pattern for `linear_with_input_q_dq_qweight_dq_output_fp32` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear_with_input_q_dq_qweight_dq

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_output_fp32
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_relu_output_fp32

corresponding pattern for `linear_with_input_q_dq_qweight_dq_output_fp32` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear_with_input_q_dq_qweight_dq

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jerryzh168 commented:
@pytorchbot merge -f "test error is unrelated to the PR"

jerryzh168 committed:
Update on "[quant] Add fused "q - qlinear - dq" operator with skipped quant op for output of linear"


Summary:
Added two ops:
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_output_fp32
* torch.ops.quantized.linear_with_input_q_dq_qweight_dq_relu_output_fp32

corresponding pattern for `linear_with_input_q_dq_qweight_dq_output_fp32` would be:
```
input -> q* -> dq* -> linear* ->
           qweight -> dq* /
```

Test Plan:
python test/test_quantization.py -k TestQuantizedLinear.test_qlinear_with_input_q_dq_qweight_dq

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
